File: 1807-evaluate-the-bracket-pairs-of-a-string/1807-evaluate-the-bracket-pairs-of-a-string.js
/**
 * @param {string} s
 * @param {string[][]} knowledge
 * @return {string}
 */
var evaluate = function(s, knowledge) {
    const map = new Map();

    // store key-value pairs
    for (let i = 0; i < knowledge.length; i++) {
        map.set(knowledge[i][0], knowledge[i][1]);
    }

    const result = [];
    let i = 0;

    while (i < s.length) {
        // normal character
        if (s[i] !== '(') {
            result.push(s[i]);
            i++;
            continue;
        }

        // find closing bracket
        let j = i + 1;
        while (s[j] !== ')') {
            j++;
        }

        // extract key
        const key = s.substring(i + 1, j);

        // replace with value or '?'
        result.push(map.has(key) ? map.get(key) : "?");

        // move after ')'
        i = j + 1;
    }

    return result.join('');
};

File: 1807-evaluate-the-bracket-pairs-of-a-string/1807-evaluate-the-bracket-pairs-of-a-string.java
class Solution {
    public String evaluate(String s, List<List<String>> knowledge) {
        
        Map<String, String> map = new HashMap<>();

        // store key-value pairs
        for (List<String> pair : knowledge) map.put(pair.get(0), pair.get(1));

        StringBuilder result = new StringBuilder();

        int i = 0;

        while (i < s.length()) {

            // normal character
            if (s.charAt(i) != '(') {
                result.append(s.charAt(i));
                i++;
                continue;
            }

            // find closing bracket
            int j = i + 1;
            while (s.charAt(j) != ')') j++;

            // extract key
            String key = s.substring(i + 1, j);

            // replace with value or '?'
            result.append(map.getOrDefault(key, "?"));

            // move after ')'
            i = j + 1;
        }

        return result.toString();
    }
}